liquidity/liquidity.go

```diff
@@ -1,21 +1,112 @@
 // Package liquidity is responsible for monitoring our node's liquidity. It
 // allows setting of a liquidity rule which describes the desired liquidity
 // balance on a per-channel basis.
+//
+// Swap suggestions are limited to channels that are not currently being used
+// for a pending swap. If we are currently processing an unrestricted swap (ie,
+// a loop out with no outgoing channel targets set or a loop in with no last
+// hop set), we will not suggest any swaps because these swaps will shift the
+// balances of our channels in ways we can't predict.
+//
+// The suggestions provided are also limited by fee estimates per fee category
+// charged for the swap (server fee, on chain fee and off-chain routing fees).
+// Note that the off-chain fees are estimated independent of route, and may vary
+// practise. The maximum fee is calculated as follows:
+// (swap amount * server PPM/1e6) + miner fee + (swap amount * routing PPM/1e6)
+// + (prepay amount * prepay PPM/1e6).
 package liquidity
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"strings"
 	"sync"
+	"time"
 
+	"github.com/btcsuite/btcutil"
 	"github.com/lightninglabs/lndclient"
+	"github.com/lightninglabs/loop"
+	"github.com/lightninglabs/loop/loopdb"
+	"github.com/lightningnetwork/lnd/clock"
+	"github.com/lightningnetwork/lnd/lnwallet/chainfee"
 	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/routing/route"
+)
+
+const (
+	// defaultFailureBackoff is the default amount of time we backoff if
+	// a channel is part of a temporarily failed swap.
+	defaultFailureBackoff = time.Hour * 24
+
+	// FeeBase is the base that we use to express fees.
+	FeeBase = 1e6
+
+	// defaultSwapFeePPM is the default limit we place on swap fees,
+	// expressed as parts per million of swap volume, 0.5%.
+	defaultSwapFeePPM = 5000
+
+	// defaultRoutingFeePPM is the default limit we place on routing fees
+	// for the swap invoice, expressed as parts per million of swap volume,
+	// 1%.
+	defaultRoutingFeePPM = 10000
+
+	// defaultRoutingFeePPM is the default limit we place on routing fees
+	// for the prepay invoice, expressed as parts per million of prepay
+	// volume, 0.5%.
+	defaultPrepayRoutingFeePPM = 5000
+
+	// defaultMaximumMinerFee is the default limit we place on miner fees
+	// per swap. We apply a multiplier to this default fee to guard against
+	// the case where we have broadcast the preimage, then fees spike and
+	// we need to sweep the preimage.
+	defaultMaximumMinerFee = 15000 * 100
+
+	// defaultMaximumPrepay is the default limit we place on prepay
+	// invoices.
+	defaultMaximumPrepay = 30000
+
+	// defaultSweepFeeLimit is the default limit we place on estimated
+	// sweep fees, (750 *4 /1000 = 3 sat/byte).
+	defaultSweepFeeLimit = chainfee.SatPerKWeight(750)
```

Comment:
nit: call this `defaultSweepFeeRateLimit` to distinguish from absolute fee limits like `defaultMaximumMinerFee`?
Also, it's `sat/vByte` here as well I think.